configuration/configuration.go

```diff
@@ -0,0 +1,111 @@
+package configuration
+
+import (
+	"errors"
+	"fmt"
+	"github.com/01org/ciao/payloads"
+	"gopkg.in/yaml.v2"
+	"net/url"
+)
+
+// we can have values set to default, except for
+//    scheduler { storage_uri }
+//    controller { compute_ca, compute_cert, identity_user, identity_password }
+//    launcher { compute_net, mgmt_net }
+//    image_service { url }
+//    identity_service { url }
+//
+// so we need to have at least those values set in our config
+//
+// TODO: proper validation of values set in yaml setup
+func validMinConf(conf *payloads.Configure) bool {
+	return (conf.Configure.Scheduler.ConfigStorageURI != "" &&
+		conf.Configure.Controller.ComputeCACert != "" &&
+		conf.Configure.Controller.ComputeCert != "" &&
+		conf.Configure.Controller.IdentityUser != "" &&
+		conf.Configure.Controller.IdentityPassword != "" &&
+		conf.Configure.Launcher.ComputeNetwork != "" &&
+		conf.Configure.Launcher.ManagementNetwork != "" &&
+		conf.Configure.ImageService.URL != "" &&
+		conf.Configure.IdentityService.URL != "")
+}
+
+func getConfiguration(yamlConf []byte) (conf payloads.Configure, err error) {
+	err = yaml.Unmarshal(yamlConf, &conf)
+	if err != nil {
+		return conf, err
+	}
+	return
+}
+
+func getPayload(conf *payloads.Configure) (payload []byte, err error) {
+	if !validMinConf(conf) {
+		return nil, errors.New(
+			"minimal configuration is not met or yaml is malformed")
+	}
+	payload, err = yaml.Marshal(&conf)
+	if err != nil {
+		return nil, err
+	}
+	return payload, nil
+}
+
+func fillDefaults(conf *payloads.Configure) {
+	if conf.Configure.Scheduler.ConfigStorageType == "" {
+		conf.Configure.Scheduler.ConfigStorageType = payloads.Filesystem
+	}
+	if conf.Configure.Controller.ComputePort == 0 {
+		conf.Configure.Controller.ComputePort = 8774
+	}
+	if conf.Configure.ImageService.Type == "" {
+		conf.Configure.ImageService.Type = payloads.Glance
+	}
+	if conf.Configure.IdentityService.Type == "" {
+		conf.Configure.IdentityService.Type = payloads.Keystone
+	}
+}
+
+// TODO: add etcd support related scheme(s)
+func discoverDriver(uriStr string) (storageType payloads.StorageType, err error) {
+	uri, err := url.Parse(uriStr)
+	if err != nil {
+		return storageType, err
+	}
+	if uri.Scheme != "file" {
+		return storageType, fmt.Errorf(
+			"Configuration URI Scheme '%s' not supported", uri.Scheme)
+	}
+	storageType = payloads.Filesystem
+	return storageType, nil
+}
+
+// GetConfiguration returns the payload to be set
+// in the SSNTP server. It needs an URI string
+// where it can read the configuration in yaml format.
+// If the URI is not valid or the yaml is either incomplete or
+// malformed, the payload recieved will be nil and err should
+// specify what the failure is
+func GetConfiguration(uri string) (payload []byte, err error) {
```

Comment:
Given the description you wrote, I guess you mean to the `getConfiguration` function rather than `getPayload`?

**`getConfiguration`** takes a configuration blob  and returns a payloads.Configure struct, while **`getPayload`** is quite the opposite; takes a payloads.Configure struct and returns a configuration blob `(payload []byte)`